### ISSUE TITLE
RFC: Provide a declarative syntax 

### ISSUE THREAD

**Issue by mandel-macaque** [sharpliner/sharpliner#9]
While a fluent syntax was very nice before we had collection initialisers, property initialisers and other new syntax in C# nowadays we can (at the same time) provide a declarative syntax with the  fluent one. The way I see it is as follows:

## Declarative approach

```csharp
// declare a pipeline with either jobs add stages
Pipeline {
    Name =  "Example pipeline",
    Resources = [
        new Repository {Url = ""},
    ],
    Variables = [
        new Variable<String> {Name = "Env", Value = "String" },
        new Variable<bool> {Name ="RunTests", Value = true },
    ],
    Stages = {
        new Stage {
            Variables = [
                new Variable<String> {Name = "StageEnv", Value = "String" },
            ],
            new Job {
                DisplayName: "My naive job",
                Steps = {
                    new Bash {
                        Name = "Naive bash step",
                        Path = "test-example.sh",  
                    }
               }   
           }
       }
    }
}
```

With the above, we can take advance of several language features, specially from C# 9 that will make it nice to work with AND intellisense will work. 

* Object property initializer
* Collection initiallizers
* Variables can use generics, we should be able to infer the type and not even need to add the type.

We will need some validation, for example for loops etc... but we can make our users write very straight forward declarative definitions that are type safe.

**Comment by mandel-macaque** [comment]
@premun look at the draft PR attached, I have made the code much cleaner to write:
```csharp
var st = new Stage
{
    Name = "MyName",
    DisplayName = "VisibleName",
    Variables = {
        new ("", ""),
        new ("", "")
    },
    Jobs = {
        new ()
        {
            Name = "First Job",
            DisplayName = "My DisplayName",
            Steps = {
            }
        },
    },
};
```

**Comment by premun** [comment]
I like it, but I also like that for ctors, you can specify the required arguments.

With the current way, you can do a very similar thing though:
```
var st = new Stage
(
    Name: "MyName",
    DisplayName: "VisibleName",
    Variables: {
        new ("", ""),
        new ("", "")
    },
    Jobs: [
        new (
            Name: "First Job",
            DisplayName: "My DisplayName",
            Steps: {
            }
        ),
    ],
);
```
Just swap `{}` with `()`, `=` with `:`. But I understand it's less csharpey..

But it would be easier to write validation (e.g. stage name is only `a-zA-Z0-9`).
Without ctor, you will have to have backing fields and validate in the `init`.

How about we do a mix? We do required params in ctor and the rest as `init`?

**Comment by mandel-macaque** [comment]
I was precisely thinking about the validation of the model once I wrote this. Lest think if we have a mix:
```csharp
var st = new Stage
(
    Name: "MyName",
    DisplayName: "DIsplayName"
)
{
    Variables = {
        new ("", ""),
        new ("", "")
    },
    Jobs = {
        new ()
        {
            Name = "First Job",
            DisplayName = "My DisplayName",
            Steps = {
            }
        },
    },
};
```

I bet that the usage of () and {} for very similar things is going to confuse the hell out of users and we are back to the horrible horrible problem of ${, $[ in yaml. I do have a proposal on how to fix this, provide a Roslyn analyser that will do the validation on the fly. YET I think there is value in constructors for special types in the pipeline. Variables, Properties, Resources and others that are small, should use constructors. Why? if you look at the code in the Variables list, it is so much nicer to do:

```csharp
Variables: [
        new ("", ""),
        new ("", ""),
],
```
than
```csharp
Variables: [
        new {Name = "FOO", Value = "BAR"}
        new (Name = "TEST", Value = ""),
    ],
```

So I think we need both, but not for validation but for API design. We work on a declarative syntax that is focused on elegance even when it allows you to make mistakes and then provide a Roslyn Analyser for things like:

* Property values, like you mentioned, we need validation of strings etc
* Collection validation, a stage with no jobs makes no sense.
* Reference checks, for example, dependsOn = "MyJob" when "MyJob" does not exist.

The way I am thinking we could have the following for our users workflow

1. Open our own special template of project in VS, contains ref to nugets, Roslyn analyser etc
2. User write pipeline definition in C# using Intellisense + analyser.
3. Compiles

Regarding 3 I have one  idea, how do I get the yaml to output? My idea, we provide the command line of the project as the output of it, we can do that by telling the user to define the Pipeline with a top level statement:
```csharp
vr pipeline = new Pipeline {
   Name = "",
  // etc
}
```
We have a code generator, that will take that top level statement and generate all the stuff needed for the cmd, so add something like:
```csharp
var cli = new SharpLinerCLI (pipeline);
cli.Execute();
```
The analyser will make sure that the user just has one pipeline. Then they can use the command line to generate it:
```bash
your.exe --generate --ouput mypipeline.yml 
```

The Roslyn analyser and the CLI can share the logic for the validation, so if the user made a mistake because he ignored the analyser, he will get the errors when calling the cmd. The code generator is just to add even more magic to the project, first versions can make you write that. We also need to make the usage of the analyser shareable. That is, do not write a validation per class, lets use reflexion and attributes, for example:

```cshap
public record Stage
{
         [Validation(RegEx("a-zA-Z0-9"))]
         public string Name { get; init; }

          [Validation(RegEx("a-zA-Z0-9"))]
         public string DisplayName { get; init; }
```

This will allow people that share dlls with new steps, jobs, etc.. to add some checks to be picked up by the analyser.

In summary, I think we can make sure people don't make mistake with good docs for the tooltips in intellisense + roslyn analysers. I think we can get the basic model without conditions (later on that) fast and that most of the work will go into the validation and the analyser.

**Comment by premun** [comment]
1. You are right, we can always have the validation during serialization. We don't need to do that in constructors.
2. We should still let the users know which arguments are required otherwise when they would be new to some build task, they would need to add one, f5, check for errors, ... I think the static analysis which can happen when the user is writing the pipeline should cover the required fields where missing fields is invalid syntax basically.
3. I think, since we aim to have a declarative syntax, we could let the user add attributes to a `Pipeline` definition and scan DLLs without the user needing to add somewhere in their workflow the actual serialization:
```csharp
[PipelineDefinition("azure-pipelines.ci.yml")]
public static readonly Pipeline CiPipeline = new {
    ...
};

[PipelineDefinition("azure-pipelines.pr.yml")]
public static readonly Pipeline PrPipeline = new {
    ...
};
```
And I guess we can have a NuGet that would hook onto the Build target and publish those files. I think these attributes might also work well with the workflow in #3.

**Comment by mandel-macaque** [comment]
Reflection for the provided dll form the user. I think we are in a good path, the analyzers are going to be fun to write 😁

**Comment by premun** [comment]
@mandel-macaque I have taken a sample YAML pipeline and I created a C# version of it.

I am not 100% happy with everything but I reckon we should somehow put it in git here or in a description and work on it until we're happy.

https://gist.github.com/premun/8693b3f0a049eaf1c35d83a79861c173

**Comment by mandel-macaque** [comment]
@premun can we add that to the wiki so that we both can edit? I have some ideas to make it nicer looking, would be a good way to keep track of changes (or maybe create the gist in the org if that is possible).

**Comment by premun** [comment]
We can edit the description of this issue? It also has a history:
https://github.com/sharpliner/sharpliner/issues/11

**Comment by premun** [comment]
I think we can close this issue?

**Comment by mandel-macaque** [comment]
@premun yes, closing
